### PR TITLE
Split native method price into Cpu and Storage fees

### LINKED ISSUE #1813: Implement cpu/storage fee split for native contracts
And adjust fee values, see neo-project/neo#2331.

### PR DESCRIPTION
**Pull request by fyrchik**
Close #1813.

### CODE REVIEW

File: pkg/interop/native/oracle/oracle.go [hunk 1]
```diff
@@ -14,3 +14,13 @@ func Request(url string, filter []byte, cb string, userData interface{}, gasForR
 		contract.States|contract.AllowNotify,
 		url, filter, cb, userData, gasForResponse)
 }
+
+// GetPrice represents `getPrice` method of NEO native contract.
```
Comment (roman-khimov) [inline pkg/interop/native/oracle/oracle.go:18]: NEO?

**Review by roman-khimov — CHANGES_REQUESTED** [review]

File: pkg/core/native/notary.go [hunk 2]
```diff
@@ -60,29 +60,29 @@ func newNotary() *Notary {
...
 	desc = newDescriptor("balanceOf", smartcontract.IntegerType,
 		manifest.NewParameter("addr", smartcontract.Hash160Type))
-	md = newMethodAndPrice(n.balanceOf, 100_0000, callflag.ReadStates)
+	md = newMethodAndPrice(n.balanceOf, 1<<15, callflag.ReadStates)
 	n.AddMethod(md, desc)
 
 	desc = newDescriptor("expirationOf", smartcontract.IntegerType,
 		manifest.NewParameter("addr", smartcontract.Hash160Type))
-	md = newMethodAndPrice(n.expirationOf, 100_0000, callflag.ReadStates)
+	md = newMethodAndPrice(n.expirationOf, 1<<15, callflag.ReadStates)
```
Comment (AnnaShaleva) [inline pkg/core/native/notary.go:85]: `getMaxNotValidBeforeDelta` and `setMaxNotValidBeforeDelta` should also be adjusted.

File: pkg/core/native/native_neo.go [hunk 3]
```diff
@@ -195,7 +212,13 @@ func (n *NEO) Initialize(ic *interop.Context) error {
 	if err != nil {
 		return err
 	}
-
+	price := big.NewInt(DefaultRegisterPrice)
+	si := state.StorageItem(bigint.ToBytes(price))
```
Comment (AnnaShaleva) [inline pkg/core/native/native_neo.go:216]: We have `setIntWithKey` helper for this in `util.go`.

File: pkg/core/native/native_neo.go [hunk 4]
```diff
@@ -323,6 +346,11 @@ func (n *NEO) PostPersist(ic *interop.Context) error {
 		n.gasPerBlock.Store(gr)
 		n.gasPerBlockChanged.Store(false)
 	}
+	if n.registerPriceChanged.Load() {
+		p := n.getRegisterPriceInternal(ic.DAO)
```
Comment (AnnaShaleva) [inline pkg/core/native/native_neo.go:350]: `n.getRegisterPriceInternal` also checks for `n.registerPriceChanged.Load`.

File: pkg/core/native/native_neo.go [hunk 5]
```diff
@@ -482,6 +510,35 @@ func (n *NEO) SetGASPerBlock(ic *interop.Context, index uint32, gas *big.Int) er
...
 
+func (n *NEO) getRegisterPrice(ic *interop.Context, _ []stackitem.Item) stackitem.Item {
+	return stackitem.NewBigInteger(big.NewInt(n.getRegisterPriceInternal(ic.DAO)))
+}
+
+func (n *NEO) getRegisterPriceInternal(d dao.DAO) int64 {
+	if !n.registerPriceChanged.Load() {
+		return n.registerPrice.Load()
+	}
+	si := d.GetStorageItem(n.ID, []byte{prefixRegisterPrice})
```
Comment (AnnaShaleva) [inline pkg/core/native/native_neo.go:521]: And we have `getIntWithKey` for this in `util.go`.

File: pkg/core/native/native_neo.go [hunk 6]
```diff
@@ -592,6 +649,7 @@ func (n *NEO) registerCandidate(ic *interop.Context, args []stackitem.Item) stac
 	} else if !ok {
 		return stackitem.NewBool(false)
 	}
+	ic.VM.AddGas(n.getRegisterPriceInternal(ic.DAO))
```
Comment (AnnaShaleva) [inline pkg/core/native/native_neo.go:652]:
```
if !ic.VM.AddGas(...) {
        panic(...)
}
```
?

File: pkg/core/native/native_neo.go [hunk 7]
```diff
@@ -482,6 +510,35 @@ func (n *NEO) SetGASPerBlock(ic *interop.Context, index uint32, gas *big.Int) er
...
+func (n *NEO) setRegisterPrice(ic *interop.Context, args []stackitem.Item) stackitem.Item {
+	price := toBigInt(args[0])
+	if price.Sign() <= 0 || !price.IsInt64() {
+		panic("invalid register price")
+	}
+	if !n.checkCommittee(ic) {
+		panic("invalid committee signature")
+	}
+	si := state.StorageItem(bigint.ToBytes(price))
+	if err := ic.DAO.PutStorageItem(n.ID, []byte{prefixRegisterPrice}, si); err != nil {
```
Comment (AnnaShaleva) [inline pkg/core/native/native_neo.go:535]: `setIntWithKey`

**Review by AnnaShaleva — CHANGES_REQUESTED** [review]

File: pkg/core/native/native_neo.go [hunk 8]
```diff
@@ -25,6 +25,7 @@ import (
 	"github.com/nspcc-dev/neo-go/pkg/smartcontract/manifest"
 	"github.com/nspcc-dev/neo-go/pkg/util"
 	"github.com/nspcc-dev/neo-go/pkg/vm/stackitem"
+	uatomic "go.uber.org/atomic"
```
[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/core/native/native_neo.go:28]: Fixed.

File: pkg/interop/native/oracle/oracle.go (hunk 1 above)
Comment (fyrchik) [reply pkg/interop/native/oracle/oracle.go:18]: Fixed.

File: pkg/core/native/notary.go (hunk 2 above)
Comment (fyrchik) [reply pkg/core/native/notary.go:85]: Fixed.

File: pkg/core/native/native_neo.go (hunk 3 above)
Comment (fyrchik) [reply pkg/core/native/native_neo.go:216]: Fixed.

File: pkg/core/native/native_neo.go (hunk 6 above)
Comment (fyrchik) [reply pkg/core/native/native_neo.go:652]: Yeah, missed that.

File: pkg/core/native/native_neo.go (hunk 4 above)
Comment (fyrchik) [reply pkg/core/native/native_neo.go:350]: Fixed.

File: pkg/core/native/native_neo.go (hunk 7 above)
Comment (fyrchik) [reply pkg/core/native/native_neo.go:535]: Fixed.

File: pkg/vm/emit/emit.go [hunk 9]
```diff
@@ -59,14 +59,18 @@ func Int(w *io.BinWriter, i int64) {
...
-		Opcodes(w, opcode.PUSHINT8+opcode.Opcode(padSize))
-		w.WriteBytes(padRight(1<<padSize, buf))
+		bigInt(w, big.NewInt(i))
 	}
 }
 
+func bigInt(w *io.BinWriter, n *big.Int) {
+	buf := bigint.ToPreallocatedBytes(n, make([]byte, 0, 32))
+	// l != 0 becase of switch
+	padSize := byte(8 - bits.LeadingZeros8(byte(len(buf)-1)))
```
Comment (AnnaShaleva) [inline pkg/vm/emit/emit.go:69]: We now use it outside of the `switch` as a separate function, so `// l != 0 becase of switch` comment is not relevant. Is it OK to have padSize = 0 here?

Comment (roman-khimov) [reply pkg/vm/emit/emit.go:69]: The problem is that `len(buf)` can be `0` for `n == 0`, so yes, it is a problem.

Comment (roman-khimov) [reply pkg/vm/emit/emit.go:69]: Push?

**Comment by roman-khimov** [comment]
Needs to be updated after #1824.

**Review by roman-khimov — APPROVED** [review]

**Review by AnnaShaleva — APPROVED** [review]

**Pull request merged**